File: dotnet/Serpent/ObjectifyVisitor.cs
/// <summary>
/// Serpent, a Python literal expression serializer/deserializer
/// (a.k.a. Python's ast.literal_eval in .NET)
///
/// Copyright 2014, Irmen de Jong (irmen@razorvine.net)
/// Software license: "MIT software license". See http://opensource.org/licenses/MIT
/// </summary>

using System;
using System.Collections.Generic;
using System.Collections;

namespace Razorvine.Serpent
{
	/// <summary>
	/// Ast nodevisitor that turns the AST into actual .NET objects (array, int, IDictionary, string, etc...)
	/// </summary>
	public class ObjectifyVisitor: Ast.INodeVisitor
	{
		private Stack<object> generated = new Stack<object>();
		private Func<IDictionary, object> dictToInstance = null;

		/// <summary>
		/// Create the visitor that converts AST in actual objects.
		/// </summary>
		public ObjectifyVisitor()
		{
		}
		
		/// <summary>
		/// Create the visitor that converts AST in actual objects.
		/// </summary>
		/// <param name="dictToInstance">functin to convert dicts to actual instances for a class,
		/// instead of leaving them as dictionaries. Requires the __class__ key to be present
		/// in the dict node. If it returns null, the normal processing is done.</param>
		public ObjectifyVisitor(Func<IDictionary, object> dictToInstance)
		{
			this.dictToInstance = dictToInstance;
		}

		/// <summary>
		/// get the resulting object tree.
		/// </summary>
		public object GetObject()
		{
			return generated.Pop();
		}
		
		public void Visit(Ast.ComplexNumberNode complex)
		{
			generated.Push(new ComplexNumber(complex.Real, complex.Imaginary));
		}
		
		public void Visit(Ast.DictNode dict)
		{
			IDictionary obj = new Hashtable(dict.Elements.Count);
			foreach(Ast.KeyValueNode kv in dict.Elements)
			{
				kv.Key.Accept(this);
				object key = generated.Pop();
				kv.Value.Accept(this);
				object value = generated.Pop();
				obj[key] = value;
			}

			if(dictToInstance==null || !obj.Contains("__class__"))
			{
				generated.Push(obj);
			}
			else
			{
				object result = dictToInstance(obj);
				if(result==null)
					generated.Push(obj);
				else
					generated.Push(result);
			}
		}
		
		public void Visit(Ast.ListNode list)
		{
			IList<object> obj = new List<object>(list.Elements.Count);
			foreach(Ast.INode node in list.Elements)
			{
				node.Accept(this);
				obj.Add(generated.Pop());
			}
			generated.Push(obj);
		}
		
		public void Visit(Ast.NoneNode none)
		{
			generated.Push(null);
		}
		
		public void Visit(Ast.IntegerNode value)
		{
			generated.Push(value.Value);
		}
		
		public void Visit(Ast.LongNode value)
		{
			generated.Push(value.Value);
		}
		
		public void Visit(Ast.DoubleNode value)
		{
			generated.Push(value.Value);
		}
		
		public void Visit(Ast.BooleanNode value)
		{
			generated.Push(value.Value);
		}
		
		public void Visit(Ast.StringNode value)
		{
			generated.Push(value.Value);
		}
		
		public void Visit(Ast.DecimalNode value)
		{
			generated.Push(value.Value);
		}
		
		public void Visit(Ast.SetNode setnode)
		{
			HashSet<object> obj = new HashSet<object>();
			foreach(Ast.INode node in setnode.Elements)
			{
				node.Accept(this);
				obj.Add(generated.Pop());
			}
			generated.Push(obj);
		}
		
		public void Visit(Ast.TupleNode tuple)
		{
			object[] array = new object[tuple.Elements.Count];
			int index=0;
			foreach(Ast.INode node in tuple.Elements)
			{
				node.Accept(this);
				array[index++] = generated.Pop();
			}
			generated.Push(array);
		}
	}
}

File: java/src/main/java/net/razorvine/serpent/ObjectifyVisitor.java
/**
 * Serpent, a Python literal expression serializer/deserializer
 * (a.k.a. Python's ast.literal_eval in Java)
 * Software license: "MIT software license". See http://opensource.org/licenses/MIT
 * @author Irmen de Jong (irmen@razorvine.net)
 */

package net.razorvine.serpent;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;
import net.razorvine.serpent.ast.*;

/**
 * Ast nodevisitor that turns the AST into actual Java objects (array, int, IDictionary, string, etc...)
 */
public class ObjectifyVisitor implements INodeVisitor
{
	Stack<Object> generated = new Stack<Object>();
	protected IDictToInstance dictConverter = null;
	
	public ObjectifyVisitor() {
	}
	
	public ObjectifyVisitor(IDictToInstance dictConverter) {
		this.dictConverter = dictConverter;
	}

	/**
	 * get the resulting object tree.
	 */
	public Object getObject()
	{
		return generated.pop();
	}

	public void visit(ComplexNumberNode complex)
	{
		generated.push(new ComplexNumber(complex.real, complex.imaginary));
	}

	public void visit(DictNode dict)
	{
		Map<Object, Object> obj = new HashMap<Object, Object>(dict.elements.size());
		for(INode e: dict.elements)
		{
			KeyValueNode kv = (KeyValueNode)e;
			kv.key.accept(this);
			Object key = generated.pop();
			kv.value.accept(this);
			Object value = generated.pop();
			obj.put(key, value);
		}

		if(dictConverter==null || !obj.containsKey("__class__"))
		{
			generated.push(obj);
		}
		else
		{
			Object result;
			try {
				result = dictConverter.convert(obj);
			} catch (IOException e) {
				throw new RuntimeException("problem converting dict to class", e);
			}
			if(result==null)
				generated.push(obj);
			else
				generated.push(result);
		}
	}

	public void visit(ListNode list)
	{
		List<Object> obj = new ArrayList<Object>(list.elements.size());
		for(INode node: list.elements)
		{
			node.accept(this);
			obj.add(generated.pop());
		}
		generated.push(obj);
	}

	public void visit(NoneNode none)
	{
		generated.push(null);
	}

	public void visit(IntegerNode value)
	{
		generated.push(value.value);
	}

	public void visit(LongNode value)
	{
		generated.push(value.value);
	}

	public void visit(DoubleNode value)
	{
		generated.push(value.value);
	}

	public void visit(BooleanNode value)
	{
		generated.push(value.value);
	}

	public void visit(StringNode value)
	{
		generated.push(value.value);
	}

	public void visit(BigIntNode value)
	{
		generated.push(value.value);
	}

	public void visit(SetNode setnode)
	{
		Set<Object> obj = new HashSet<Object>();
		for(INode node: setnode.elements)
		{
			node.accept(this);
			obj.add(generated.pop());
		}
		generated.push(obj);
	}

	public void visit(TupleNode tuple)
	{
		Object[] array = new Object[tuple.elements.size()];
		int index=0;
		for(INode node: tuple.elements)
		{
			node.accept(this);
			array[index++] = generated.pop();
		}
		generated.push(array);
	}
	
}
